### PR TITLE
shutdown 10 seconds after os.signal

### PR DESCRIPTION
**Pull request by ramon-reichert**
Like I left commented, we should think if 10 seconds is the right time to wait.
The code is working, but I think we should test it together with real interruptions and error simulations.

### CODE REVIEW

File: cmd/api/main.go [hunk 1]
```diff
@@ -49,9 +53,23 @@ func run() error {
...
-	err = server.ListenAndServe()
-	if err != nil && !errors.Is(err, http.ErrServerClosed) {
-		return fmt.Errorf("unexpected http server error: %w", err)
+	go func() (err error) {
+		err = server.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			return fmt.Errorf("unexpected http server error: %w", err)
+		}
+		return nil
+	}()
```
Comment (cesarFuhr) [inline cmd/api/main.go]: I don't think this function should be returning an error. You never read it.
```suggestion
	go func() {
		err = server.ListenAndServe()
		if err != nil && !errors.Is(err, http.ErrServerClosed) {
			log.Println("unexpected http server error: %v", err)
		}
		log.Println("stopped serving new requests.")
	}()

File: cmd/api/main.go [hunk 2]
```diff
@@ -49,9 +53,23 @@ func run() error {
...
+
+	sc := make(chan os.Signal, 1)
+	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM)
+	<-sc
+
+	ctx, shutdownRelease := context.WithTimeout(context.Background(), 10*time.Second) //HOW MUCH TIME MAKE SENSE FOR US??
+	defer shutdownRelease()
```
Comment (cesarFuhr) [inline cmd/api/main.go]: I don't like the cancel function name here (`shutdownRelease`)... I would stick with the usual `cancel` or go for something like `shutdownCancel`.

**Review by cesarFuhr — CHANGES_REQUESTED** [review]

**Comment by cesarFuhr** [comment]
Nice job, I would like to  talk to you about this.
1 - How do you tested it?
2 - Do you know how this actually works?
3 - Have you done research around `channels`and `goroutines`?
4 - Does this code has concurrent execution?

File: cmd/api/main.go [hunk 3]
```diff
@@ -49,9 +53,23 @@ func run() error {
...
+			return fmt.Errorf("unexpected http server error: %w", err)
+		}
+		return nil
+	}()
+
+	sc := make(chan os.Signal, 1)
+	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM)
+	<-sc
+
+	ctx, shutdownRelease := context.WithTimeout(context.Background(), 10*time.Second) //HOW MUCH TIME MAKE SENSE FOR US??
```
Comment (cesarFuhr) [inline cmd/api/main.go]: This timeout value should be configurable and we should read it from an environment variable. We also should have an default value, that could be 10 seconds.

Comment (ramon-reichert) [reply cmd/api/main.go]: Ok. We can talk this afternoon if you want. Just tell me at whatsapp.

File: cmd/api/main.go [hunk 4]
```diff
@@ -49,9 +54,32 @@ func run() error {
...
+		}
+		log.Println("stopped serving new requests.")
+	}()
+
+	sc := make(chan os.Signal, 1)
+	signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM)
+	<-sc
+
+	deadline := 10
+	deadlineStr := os.Getenv("DEADLINE")
```
Comment (cesarFuhr) [inline cmd/api/main.go]: Lets go with `SERVICE_SHUTDOWN_TIMEOUT`.

Comment (cesarFuhr) [reply cmd/api/main.go]: Lets use the `time.Duration` type. You can use this function to parse it https://pkg.go.dev/time#ParseDuration .

File: cmd/api/main.go [hunk 5]
```diff
@@ -49,9 +54,32 @@ func run() error {
...
+	if deadlineStr != "" {
+		deadline, err = strconv.Atoi(deadlineStr)
+		if err != nil {
+			return fmt.Errorf("getting deadline from env: %w", err)
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(deadline)*time.Second)
```
Comment (cesarFuhr) [inline cmd/api/main.go]: If you use duration you wont need to multiply it by seconds here.

File: docker-compose.yml [hunk 6]
```diff
@@ -14,6 +14,7 @@ services:
     environment:
       DATABASE_URL: "postgres://postgres:chevas@db:5432/booksdb?sslmode=disable"
       DATABASE_MIGRATIONS_PATH: "/src/migrations"
+      DEADLINE: "10"
```
Comment (cesarFuhr) [inline docker-compose.yml]: Don't forget to change it here.
You can see why I am asking you to change the env variable name by looking at this env declaration in the docker-compose file here. What deadline stands for? You cannot tell looking from the outside. You would have to inspect the code to make sure what it actually represents.

**Review by cesarFuhr — CHANGES_REQUESTED** [review]

**Review by cesarFuhr — APPROVED** [review]

**Pull request merged**